File: src/integrationTest/java/com/pubnub/api/integration/PushIntegrationTest.java
package com.pubnub.api.integration;

import com.pubnub.api.PubNubException;
import com.pubnub.api.enums.PNPushEnvironment;
import com.pubnub.api.enums.PNPushType;
import com.pubnub.api.integration.util.BaseIntegrationTest;
import com.pubnub.api.models.consumer.push.PNPushListProvisionsResult;
import org.junit.Test;

import java.util.*;

import static com.pubnub.api.integration.util.RandomGenerator.randomNumber;
import static org.junit.Assert.*;

public class PushIntegrationTest extends BaseIntegrationTest {

    private final List<String> expectedChannels = new ArrayList<>();
    private String expectedDeviceId;
    private String expectedTopic;

    @Override
    protected void onBefore() {
        for (int i = 0; i < 3; i++) {
            expectedChannels.add(UUID.randomUUID().toString().substring(0, 8).toUpperCase(Locale.US));
        }

        final StringBuilder builder = new StringBuilder();
        for (int i = 0; i < 70; i++) {
            builder.append(randomNumber(0, 9));
        }
        expectedDeviceId = builder.toString();

        expectedTopic = UUID.randomUUID().toString();
    }

    @SuppressWarnings("deprecation")
    @Test
    public void testEnumNames() {
        assertEquals("apns", PNPushType.APNS.toString());
        assertEquals("gcm", PNPushType.GCM.toString());
        assertEquals("gcm", PNPushType.FCM.toString());
        assertEquals("mpns", PNPushType.MPNS.toString());
        assertEquals("apns2", PNPushType.APNS2.toString());
    }

    @Test
    public void testCycle() throws PubNubException, InterruptedException {
        for (PNPushType value : PNPushType.values()) {
            performCompleteCycle(value);
        }
    }

    private void performCompleteCycle(PNPushType pushType) throws InterruptedException, PubNubException {
        pubNub.addPushNotificationsOnChannels()
                .channels(expectedChannels)
                .pushType(pushType)
                .topic(expectedTopic)
                .deviceId(expectedDeviceId)
                .sync();

        Thread.sleep(1000);

        final PNPushListProvisionsResult result = pubNub.auditPushChannelProvisions()
                .deviceId(expectedDeviceId)
                .pushType(pushType)
                .topic(expectedTopic)
                .environment(PNPushEnvironment.DEVELOPMENT)
                .sync();
        assert result != null;
        assertTrue(result.getChannels().containsAll(expectedChannels));

        Thread.sleep(1000);

        pubNub.removePushNotificationsFromChannels()
                .pushType(pushType)
                .environment(PNPushEnvironment.DEVELOPMENT)
                .deviceId(expectedDeviceId)
                .topic(expectedTopic)
                .channels(Collections.singletonList(expectedChannels.get(0)))
                .sync();

        Thread.sleep(1000);

        final PNPushListProvisionsResult oneRemoved = pubNub.auditPushChannelProvisions()
                .deviceId(expectedDeviceId)
                .pushType(pushType)
                .topic(expectedTopic)
                .environment(PNPushEnvironment.DEVELOPMENT)
                .sync();
        assert oneRemoved != null;
        assertFalse(oneRemoved.getChannels().isEmpty());
        assertFalse(oneRemoved.getChannels().contains(expectedChannels.get(0)));

        Thread.sleep(1000);

        pubNub.removeAllPushNotificationsFromDeviceWithPushToken()
                .pushType(pushType)
                .environment(PNPushEnvironment.DEVELOPMENT)
                .deviceId(expectedDeviceId)
                .topic(expectedTopic)
                .sync();

        Thread.sleep(1000);

        final PNPushListProvisionsResult listResult = pubNub.auditPushChannelProvisions()
                .deviceId(expectedDeviceId)
                .pushType(pushType)
                .topic(expectedTopic)
                .environment(PNPushEnvironment.DEVELOPMENT)
                .sync();

        assert listResult != null;
        assertTrue(listResult.getChannels().isEmpty());
        assertFalse(listResult.getChannels().containsAll(expectedChannels));
    }
}

File: pubnub-kotlin/pubnub-kotlin-impl/src/integrationTest/kotlin/com/pubnub/api/integration/PushIntegrationTest.kt
package com.pubnub.api.integration

import com.pubnub.api.enums.PNPushEnvironment
import com.pubnub.api.enums.PNPushType
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import java.util.Locale
import java.util.UUID

class PushIntegrationTest : BaseIntegrationTest() {
    lateinit var expectedChannels: List<String>
    lateinit var expectedDeviceId: String
    lateinit var expectedTopic: String

    override fun onBefore() {
        expectedChannels =
            generateSequence { UUID.randomUUID().toString().substring(0, 8).uppercase(Locale.getDefault()) }.take(3)
                .toList()
        expectedDeviceId =
            generateSequence { (0..9).random() }.take(70).toList().shuffled().joinToString(separator = "")
        expectedTopic = UUID.randomUUID().toString()
    }

    @Test
    fun testEnumNames() {
        assertEquals("apns", PNPushType.APNS.toParamString())
        assertEquals("fcm", PNPushType.FCM.toParamString())
        assertEquals("apns2", PNPushType.APNS2.toParamString())
    }

    @Test
    fun testCycle() {
        PNPushType.values().forEach {
            runPushOperations(it)
        }
    }

    private fun runPushOperations(pushType: PNPushType) {
        println("Push type '$pushType'")
        println("Channels $expectedChannels'")
        println("Topic $expectedTopic'")
        println("GeneratedToken '${expectedDeviceId.length}': $expectedDeviceId")

        pubnub.addPushNotificationsOnChannels(
            channels = expectedChannels,
            pushType = pushType,
            topic = expectedTopic,
            deviceId = expectedDeviceId,
        ).sync()

        wait()

        pubnub.auditPushChannelProvisions(
            deviceId = expectedDeviceId,
            pushType = pushType,
            topic = expectedTopic,
            environment = PNPushEnvironment.DEVELOPMENT,
        ).sync().run {
            assertTrue(channels.containsAll(expectedChannels))
        }

        wait()

        pubnub.removePushNotificationsFromChannels(
            pushType = pushType,
            environment = PNPushEnvironment.DEVELOPMENT,
            deviceId = expectedDeviceId,
            topic = expectedTopic,
            channels = listOf(expectedChannels[0]),
        ).sync()

        wait()

        pubnub.auditPushChannelProvisions(
            deviceId = expectedDeviceId,
            pushType = pushType,
            topic = expectedTopic,
            environment = PNPushEnvironment.DEVELOPMENT,
        ).sync().apply {
            assertFalse(channels.isEmpty())
            assertFalse(channels.contains(expectedChannels[0]))
        }

        wait()

        pubnub.removeAllPushNotificationsFromDeviceWithPushToken(
            pushType = pushType,
            environment = PNPushEnvironment.DEVELOPMENT,
            deviceId = expectedDeviceId,
            topic = expectedTopic,
        ).sync()

        wait()

        pubnub.auditPushChannelProvisions(
            deviceId = expectedDeviceId,
            pushType = pushType,
            topic = expectedTopic,
            environment = PNPushEnvironment.DEVELOPMENT,
        ).sync().apply {
            assertTrue(channels.isEmpty())
            assertFalse(channels.containsAll(expectedChannels))
        }
    }
}
